Adjust diagnostic messages of connection test for Connect My Computer

ravicious opened the pull request:
Closes #32206, closes #32185.

The messages had to be adjusted in two places:

* `lib/srv/authhandlers.go` – this is on the side of the server agent. Here we use the fact that `h.c.Server.GetInfo()` returns `types.Server`, so to detect a Connect My Computer node we can check if the node has `types.ConnectMyComputerNodeOwnerLabel` in its labels.
* `lib/client/conntest/ssh.go` – this is on the side of the proxy and the request to perform a connection diagnostic. We don't have access to info about the target (unless we explicitly fetch it). But at the time of making the request, we know that we're inside the Connect My Computer flow, so we pass another argument through `TestConnectionRequest`.

I also included a couple of UI improvements, each one is in its own separate commit.

| Details hidden | Details shown |
| --- | --- |
| <img width="1117" alt="conntest-no-details" src="https://github.com/gravitational/teleport/assets/27113/b5afc650-0fb9-4d9a-a6b1-540a45ea5652"> | <img width="1117" alt="conntest-details" src="https://github.com/gravitational/teleport/assets/27113/59c164d0-a4c1-4980-8361-652bdec904b5"> |

Changelog: Added a connection test when enrolling a new Connect My Computer resource in Web UI

ravicious committed:
Use JSDocs in services/agents/types.ts

ravicious committed:
Add ResourceTile to TestConnectionRequest

ravicious committed:
Adjust diagnostic messages for Connect

ravicious committed:
Remove LegacyTestConnection

ravicious committed:
TextIcon: Align icon to top rather than center

When showing multiple TextIcons one after another, like when showing
connection diagnostic traces, it makes it easier to see where one
trace starts and another ends. Especially if one of the traces spans
multiple lines.

ravicious committed:
Preserve line breaks from trace messages in UI

ravicious commented on lib/client/conntest/ssh.go:
```diff
@@ -231,10 +239,45 @@ func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDia
+					"Output from the Connect My Computer agent: %v",
+				connectMyComputerRoleName, sshPrincipal, sshPrincipal, processStdoutString,
+			)
+			// The wrapping here is done so that the detailed message will be shown under "Show details"
+			// and not as one of the main points of the connection test.
+			sshError = trace.Wrap(sshError, detailedMessage)
```
I feel like this could be considered controversial but I think it should be fine? Unless we want to explicitly include another field in the trace for those longer messages that would be shown in "Show details" along with the actual error.

kimlisa commented on lib/client/conntest/ssh.go:
```diff
@@ -231,10 +239,45 @@ func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDia
+	isInvalidNodePrincipal := isUsernameLookupFail || (isConnectMyComputerNode && isForkExecOperationNotPermitted)
+
+	if isInvalidNodePrincipal {
+		message := fmt.Sprintf("Invalid user. Please ensure the principal %q is a valid Linux login in the target node. Output from Node: %v",
```
idk if i'm just brain farting or is a dumb question... but what about non-linux login? same with systemctl, what about mac equivalent launctl? i know it was here before, and was just curious

ravicious replied:
I'll push a commit tomorrow which addresses this.

marcoandredinis commented on lib/client/conntest/connection_tester.go:
```diff
@@ -41,6 +41,11 @@ type TestConnectionRequest struct {
 	ResourceKind string `json:"resource_kind"`
 	// ResourceName is the identification of the resource's instance to test.
 	ResourceName string `json:"resource_name"`
+	// ResourceTile is an optional field which identifies the Discover tile that wants to perform a
+	// connection test.
+	// Useful in situations in which the test for the same kind of resource needs to behave
+	// differently depending on the tile which sent the request.
+	ResourceTile string `json:"resource_tile"`
```
I think we considered ResourceSubKind but that term is already in use by the Resource hierarchy, where, as you said, each resource can have multiple subkinds.
The issue is that this is not an actual subkind.
Server SubKinds are: teleport server, openssh or openssh-eice

In this case, we are talking about a Teleport Agent.
So the SubKind for this resource is already defined.
But we still need to differentiate it somehow.

We ended up with ResourceTile which has its downsides as well, but seemed to be our best bet (considering the other options).

ravicious replied:
@marcoandredinis @kimlisa Given that there are conntest users which do not depend on Discover and that in the end this PR will add three new fields to `TestConnectionRequest`, what if I use `SSHNodeKind` instead of `ResourceTile`?

It's true that `Kind` is used for many other things, but here specifically we'd talk about `SSHNodeKind` which is specific to conntest.

The PR already needs to add `SSHPrincipalSelectionMode` and the next commit to address https://github.com/gravitational/teleport/pull/35308#discussion_r1414926618 will add `SSHNodeOS`.

ravicious replied:
Ah, okay, from this angle `NodeSubKind` can indeed be confusing. `SSHNodeInstallationMode` sounds fine, though I'll shorten it to `SSHNodeSetupMethod` (`script`/`connect_my_computer`).

marcoandredinis replied:
Looks good :+1: 
Thank you for looking into this :pray:

marcoandredinis commented on lib/client/conntest/ssh.go:
```diff
@@ -231,10 +239,45 @@ func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDia
+			// address the case where the actual test gets executed in step 2.
+			message = "Invalid user."
+			detailedMessage := fmt.Sprintf(
+				"If this is step 2 of the connection test, you probably picked a login which does not match "+
+					"the system user that is running Teleport Connect with the Connect My Computer agent. "+
+					"Pick the correct login and try again.\n"+
+					"If this is step 1 of the connection test and the login was picked automatically, "+
+					"it means that the role %q includes only the login %q and %q is not a valid principal for this node. "+
+					"To fix this problem, reload this page, pick Connect My Computer again, then in Teleport Connect "+
+					"remove the Connect My Computer agent and start Connect My Computer setup again.\n"+
+					"Output from the Connect My Computer agent: %v",
```
This wall of text might be very confusing for the user
Can we somehow detect what's the actual use case?
I mean the current process's yaml config has a specific label we can check on, right?

marcoandredinis replied:
Does that information reach this section?
IIRC, I changed the certificate to include the `connection diagnostic id`, because this just a regular SSH session (so, no rpc calls)

ravicious replied:
Oh, I meant that if we add another field to `TestConnectionRequest` then we won't have to fetch the role, instead we can show the appropriate message based on the new field alone.

marcoandredinis replied:
Then I think we should do it.

ravicious replied:
Addressed in 606cc64.

ravicious committed:
Offer better diag msg depending on SSH principal selection mode

ravicious committed:
Suggest systemctl/launchctl depending on platform

ravicious committed:
Rename ResourceTile to SSHNodeSetupMethod

ravicious commented:
@kimlisa Ping.

ravicious committed:
Merge branch 'master' into ravicious/cmc-conn-test

marcoandredinis commented on lib/client/conntest/connection_tester.go:
```diff
@@ -51,6 +51,19 @@ type TestConnectionRequest struct {
+	// SSHNodeOS is an optional field which describes the OS the agent runs on.
+	// Valid values: windows, darwin, linux.
+	SSHNodeOS string `json:"ssh_node_os,omitempty"`
+	// SSHNodeSetupMethod is an optional field which describes how an SSH agent was installed.
+	SSHNodeSetupMethod string `json:"ssh_node_setup_method,omitempty"`
```
Adding Valid or Supported values would help understand a little bit which values are expected here.

ravicious committed:
Add tests for launchctl vs systemctl

ravicious committed:
Fix test after removal of "Linux" from "valid Linux login"

ravicious committed:
Add tests for SSHConnectionTester

ravicious committed:
Document valid values for SSHNodeSetupMethod

ravicious commented:
The failing test made me realize that `SSHConnectionTester` does have tests. I added some for the behavior specific to Connect My Computer.

I didn't add any for the behavior expressed in `lib/srv/authhandlers.go`. Testing those would require refactoring `webPack` to support launching a node with a specific set of static labels. `newWebPack` already accepts `opts ...proxyOption` and I couldn't figure out an easy way to make it support `lib/srv/regular.ServerOption` so that we can use `lib/srv/regular.SetLabels`.

ravicious committed:
Remove test for principal with no OS permissions to run as

In CI, it looks like we can absolutely spawn a new session as root, so
the test is meaningless there.

Instead, I'm going to add authhandlers tests and this test to the Connect
My Computer test plan.
